### ISSUE TITLE
tweakwcs.wcsimage.convex_hull() is returning NaN for seven WFISim files.

### ISSUE THREAD

**Issue by stscijgbot-rstdms** [spacetelescope/romancal#926]
_Issue [RCAL-688](https://jira.stsci.edu/browse/RCAL-688) was created on JIRA by [Mairan Teodoro](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mteodoro):_

Seven of the 36 `WFISim` files throw an error when processed through `{}TweakRegStep{`}. The files are:
 * `{}r0000501001001001001_01101_0001_WFI07_cal.asdf{`};
 * `{}r0000501001001001001_01101_0001_WFI16_cal.asdf{`};
 * `{}r0000501001001001001_01101_0001_WFI17_cal.asdf{`};
 * `{}r0000501001001001001_01101_0002_WFI01_cal.asdf{`};
 * `{}r0000501001001001001_01101_0002_WFI03_cal.asdf{`};
 * `{}r0000501001001001001_01101_0002_WFI09_cal.asdf{`};
 * `{}r0000501001001001001_01101_0002_WFI10_cal.asdf{`}.

The error is "`{}ValueError: Out of domain for acos{`}" and the traceback is:
```java
ValueError                                Traceback (most recent call last)
/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb Cell 8 line 3
     <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=22'>23</a> res = img
     <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=23'>24</a> # res = timer(DQInitStep, res)
     <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=24'>25</a> # res = timer(SaturationStep, res)
     <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=25'>26</a> # res = timer(LinearityStep, res)
   (...)
     <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=30'>31</a> # res = timer(AssignWcsStep, res)
     <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=31'>32</a> # res = timer(SourceDetectionStep, res)
---> <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=32'>33</a> res = timer(TweakRegStep, [res])
     <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=34'>35</a> # append result to list
     <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=35'>36</a> result_list.append(res)/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb Cell 8 line 6
      <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=3'>4</a> def timer(pipeline_step, input, **kwargs):
      <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=4'>5</a>     tic = datetime.datetime.now()
----> <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=5'>6</a>     res = pipeline_step.call(input, **kwargs)
      <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=6'>7</a>     toc = datetime.datetime.now()
      <a href='vscode-notebook-cell:/Users/mteodoro/ROMAN/SYNTHETIC_IMAGES/resample_synthetic_images_wfsim.ipynb#X10sZmlsZQ%3D%3D?line=7'>8</a>     dt = toc - ticFile ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/stpipe/step.py:652, in Step.call(cls, *args, **kwargs)
    649 name = config.get(\"name\", None)
    650 instance = cls.from_config_section(config, name=name, config_file=config_file)
--> 652 return instance.run(*args)File ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/stpipe/step.py:478, in Step.run(self, *args)
    476     self.prefetch(*args)
    477 try:
--> 478     step_result = self.process(*args)
    479 except TypeError as e:
    480     if \"process() takes exactly\" in str(e):File ~/ROMAN/romancal/romancal/tweakreg/tweakreg_step.py:449, in TweakRegStep.process(self, input)
    446                 del imcat.meta[\"fit_info\"]
    448         # Perform fit
--> 449         align_wcs(
    450             imcats,
    451             refcat=ref_cat,
    452             enforce_user_order=True,
    453             expand_refcat=False,
    454             minobj=self.abs_minobj,
    455             match=xyxymatch_gaia,
    456             fitgeom=self.abs_fitgeometry,
    457             nclip=self.abs_nclip,
    458             sigma=(self.abs_sigma, \"rmse\"),
    459             ref_tpwcs=imcats[0],
    460         )
    462 for imcat in imcats:
    463     image_model = imcat.meta[\"image_model\"]File ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/tweakwcs/imalign.py:557, in align_wcs(wcscat, refcat, ref_tpwcs, enforce_user_order, expand_refcat, minobj, match, fitgeom, nclip, sigma, clip_accum, group_bb_policy)
    553 if wcat.meta.get('catalog', None) is None:
    554     raise ValueError(\"Each object in 'wcscat' must have a valid \"
    555                      \"catalog.\")
--> 557 wcs_im_cat = WCSImageCatalog(
    558     catalog=wcat.meta['catalog'],
    559     corrector=wcat,
    560     name=wcat.meta.get('name', 'Unknown'),
    561     group_id=wcat.meta.get('group_id', None)
    562 )
    563 wcs_im_cat.fit_status = \"FAILED: Unknown error\"
    565 wcs_im_cats.append(wcs_im_cat)File ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/astropy/utils/decorators.py:604, in deprecated_renamed_argument.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
    601             msg += f\"\
        Use {alternative} instead.\"
    602         warnings.warn(msg, warning_type, stacklevel=2)
--> 604 return function(*args, **kwargs)File ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/tweakwcs/wcsimage.py:102, in WCSImageCatalog.__init__(self, catalog, corrector, name, group_id, meta)
     99 self._fit_info = {'status': 'SKIPPED'}
    101 self.corrector = corrector
--> 102 self.catalog = catalog
    103 self._poly_area = NoneFile ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/tweakwcs/wcsimage.py:244, in WCSImageCatalog.catalog(self, catalog)
    241     self._catalog['id'] = np.arange(1, len(self._catalog) + 1)
    243 # create spherical polygon bounding the image
--> 244 self.calc_bounding_polygon()File ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/tweakwcs/wcsimage.py:491, in WCSImageCatalog.calc_bounding_polygon(self)
    489 # create smallest convex spherical polygon bounding all sources:
    490 if self.catalog:
--> 491     self._calc_cat_convex_hull()File ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/tweakwcs/wcsimage.py:478, in WCSImageCatalog._calc_cat_convex_hull(self)
    476 self._bb_radec = (ra, dec)
    477 self._polygon = SphericalPolygon.from_radec(ra, dec)
--> 478 self._poly_area = np.fabs(self._polygon.area())File ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/spherical_geometry/polygon.py:1108, in SphericalPolygon.area(self)
   1106 area = 0.0
   1107 for subpoly in self:
-> 1108     area += subpoly.area()
   1109 return areaFile ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/spherical_geometry/polygon.py:553, in SingleSphericalPolygon.area(self)
    550     return np.array(0.0)
    552 points = np.vstack((self._points, self._points[1]))
--> 553 angles = great_circle_arc.angle(points[:-2], points[1:-1],
    554                                 points[2:], degrees=False)
    556 return np.sum(angles) - (len(angles) - 2) * np.piFile ~/.pyenv/versions/3.11.2/envs/romancal_resample_venv_dev/lib/python3.11/site-packages/spherical_geometry/great_circle_arc.py:322, in angle(A, B, C, degrees)
    298 \"\"\"
    299 Returns the angle at *B* between *AB* and *BC*.
    300 
   (...)
    319    polygon.  Graphics Gems IV.  1994.  Academic Press.
    320 \"\"\"
    321 if HAS_C_UFUNCS:
--> 322     angle = math_util.angle(A, B, C)
    323 else:
    324     A = np.asanyarray(A)ValueError: Out of domain for acos"```
 

It looks like when calling `{}WCSImageCatalog._calc_cat_convex_hull(){`}, the results from `tweakwcs.wcsimage.convex_hull()` contain `{}NaN{`}s.

For example, running `r0000501001001001001_01101_0001_WFI07_cal.asdf` through `TweakRegStep` will result in the following for (`{}RA, Dec{`}) from `{}tweakwcs.wcsimage.convex_hull(){`}:

 
```java
ra = [79.66700713 79.74057285 79.7412698 79.76786004 79.76796809 nan 79.71774705 79.71564206 79.698546 79.69628362 79.69568845 79.66874934 79.60897119 79.59113084 79.58540471 79.60440626 79.61520848 79.6168784 79.64245617 79.65131966 79.66700713]
dec = [29.84431934 29.80820374 29.8078416 29.7860489 29.78580274 nan 29.71265 29.71024942 29.69288375 29.69149943 29.69151289 29.70220816 29.73106358 29.74428309 29.75874743 29.78831562 29.80453336 29.80701569 29.844726 29.84648466 29.84431934]
```

**Comment by stscijgbot-rstdms** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/RCAL-688?focusedCommentId=666738#comment-666738):_

Where is the data?

**Comment by stscijgbot-rstdms** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/RCAL-688?focusedCommentId=667805#comment-667805):_

The error is due to the fact that source catalog passed to tweakwcs contains sources outside of the bounding box of the image's WCS. I was surprised by this because normally source finding algorithm for JWST would not find sources outside of the image. For Roman, I see that source finding algorithm was replaced with downloading catalog from a server. It is possible that the radius (SR) given to the requests.get() call is too large and returned results includes sources outside of the image.
The code in the get_catalog() function should filter returned sources and keep only the ones that are within the bounding box of the WCS.

**Comment by stscijgbot-rstdms** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/RCAL-688?focusedCommentId=667806#comment-667806):_

Bad sources by file:

*+r0000501001001001001_01101_0001_WFI07_cal.asdf:+*

282.636235462552, -1.6375547676547753

+*r0000501001001001001_01101_0001_WFI17_cal.asdf:*+

2225.7673101553632, -0.694685050029801

+*r0000501001001001001_01101_0001_WFI16_cal.asdf:*+

-1.7818295892010048, 3376.0194056854116

**Comment by stscijgbot-rstdms** [comment]
_Comment by [Eddie Schlafly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eschlafly) on [JIRA](https://jira.stsci.edu/browse/RCAL-688?focusedCommentId=667875#comment-667875):_

Thanks Mihai.  Sorry for being dense and unfamiliar with the code, but looking briefly, it looks to me like the first 'imcats' argument indeed includes catalogs derived from source detection on images, and the second 'refcat' argument is a Gaia server query.  That seems like reasonable behavior to me.  I might naively expect to want the reference catalog to go slightly beyond the nominal boundary of the image since I imagine we want to treat the WCS as preliminary at this point; is that against the rule?  Thanks!

**Comment by stscijgbot-rstdms** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/RCAL-688?focusedCommentId=668074#comment-668074):_

I think I found the bug. I will have a PR soon. The code to filter out sources outside of the bounding box is there but some code needs to be rearranged.

**Comment by stscijgbot-rstdms** [comment]
_Comment by [Mihai Cara](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mcara) on [JIRA](https://jira.stsci.edu/browse/RCAL-688?focusedCommentId=668286#comment-668286):_

A fix was proposed in #947
